pq-crypto/s2n_pq.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_pq.h"
+
+#if defined(S2N_CPUID_AVAILABLE)
+/* https://en.wikipedia.org/wiki/CPUID */
+#include <cpuid.h>
+
+/* The cpuid.h header included with older versions of gcc and
+ * clang doesn't include definitions for bit_ADX, bit_BMI2, or
+ * __get_cpuid_count(). */
+#if !defined(bit_ADX)
+#define bit_ADX 0x00080000
+#endif
+
+#if !defined(bit_BMI2)
+#define bit_BMI2 0x00000100
+#endif
+
+bool s2n_get_cpuid_count(uint32_t leaf, uint32_t sub_leaf, uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx) {
+    /* 0x80000000 probes for extended cpuid info */
+    uint32_t max_level = __get_cpuid_max(leaf & 0x80000000, 0);
+
+    if (max_level == 0 || max_level < leaf) {
+        return false;
+    }
+
+    __cpuid_count(leaf, sub_leaf, *eax, *ebx, *ecx, *edx);
+    return true;
+}
+#endif
+
+static bool sikep434r2_asm_enabled = false;
```

Comment:
I think we should move this declaration to the top of the file before any function implementations.